File: ios/Sources/Services/AchievementService.swift
import Foundation

/// Service for tracking and unlocking parental achievements and milestones
@MainActor
class AchievementService {
    static let shared = AchievementService()

    private let unlockedAchievementsKey = "unlocked_achievements"
    private var unlockedAchievements: Set<String> = []

    private init() {
        loadUnlockedAchievements()
    }

    // MARK: - Achievement Management

    /// Check for newly unlocked achievements and return them
    func checkForNewAchievements(baby: Baby, dataStore: DataStore) async -> [Achievement] {
        var newAchievements: [Achievement] = []

        for achievement in Achievement.allAchievements {
            if unlockedAchievements.contains(achievement.id) {
                continue // Already unlocked
            }

            if await achievement.isUnlocked(for: baby, dataStore: dataStore) {
                unlockedAchievements.insert(achievement.id)
                newAchievements.append(achievement)

                // Log achievement unlock
                Logger.info("Achievement unlocked: \(achievement.title)")
                await Analytics.shared.log("achievement_unlocked", parameters: [
                    "achievement_id": achievement.id,
                    "baby_id": baby.id.uuidString
                ])
            }
        }

        // Save updated achievements
        if !newAchievements.isEmpty {
            saveUnlockedAchievements()
        }

        return newAchievements
    }

    /// Get all unlocked achievements
    func getUnlockedAchievements() -> [Achievement] {
        return Achievement.allAchievements.filter { unlockedAchievements.contains($0.id) }
    }

    /// Check if achievement is unlocked
    func isAchievementUnlocked(_ achievementId: String) -> Bool {
        return unlockedAchievements.contains(achievementId)
    }

    private func loadUnlockedAchievements() {
        if let data = UserDefaults.standard.data(forKey: unlockedAchievementsKey),
           let decoded = try? JSONDecoder().decode(Set<String>.self, from: data) {
            unlockedAchievements = decoded
        }
    }

    private func saveUnlockedAchievements() {
        if let data = try? JSONEncoder().encode(unlockedAchievements) {
            UserDefaults.standard.set(data, forKey: unlockedAchievementsKey)
        }
    }
}

struct Achievement: Identifiable {
    let id: String
    let title: String
    let description: String
    let iconName: String
    let rarity: Rarity
    let unlockCondition: String
    let isUnlocked: (Baby, DataStore) async -> Bool

    enum Rarity {
        case common
        case rare
        case epic
        case legendary

        var color: String {
            switch self {
            case .common: return "gray"
            case .rare: return "blue"
            case .epic: return "purple"
            case .legendary: return "gold"
            }
        }
    }

    static let allAchievements: [Achievement] = [
        // Logging achievements
        Achievement(
            id: "first_log",
            title: "Getting Started",
            description: "Logged your first event",
            iconName: "star.fill",
            rarity: .common,
            unlockCondition: "Log your first event"
        ) { baby, dataStore in
            let events = try? await dataStore.fetchEvents(for: baby, from: Date.distantPast, to: Date.distantFuture)
            return (events?.count ?? 0) >= 1
        },

        Achievement(
            id: "hundred_logs",
            title: "Dedicated Parent",
            description: "Logged 100 events",
            iconName: "100.circle.fill",
            rarity: .rare,
            unlockCondition: "Log 100 events"
        ) { baby, dataStore in
            let events = try? await dataStore.fetchEvents(for: baby, from: Date.distantPast, to: Date.distantFuture)
            return (events?.count ?? 0) >= 100
        },

        Achievement(
            id: "thousand_logs",
            title: "Super Parent",
            description: "Logged 1,000 events",
            iconName: "1000.circle.fill",
            rarity: .epic,
            unlockCondition: "Log 1,000 events"
        ) { baby, dataStore in
            let events = try? await dataStore.fetchEvents(for: baby, from: Date.distantPast, to: Date.distantFuture)
            return (events?.count ?? 0) >= 1000
        },

        // Streak achievements
        Achievement(
            id: "week_streak",
            title: "Week Warrior",
            description: "Logged events for 7 consecutive days",
            iconName: "calendar.badge.clock",
            rarity: .rare,
            unlockCondition: "Log events for 7 days in a row"
        ) { baby, dataStore in
            await checkConsecutiveDays(baby: baby, dataStore: dataStore, requiredDays: 7)
        },

        Achievement(
            id: "month_streak",
            title: "Month Master",
            description: "Logged events for 30 consecutive days",
            iconName: "calendar.badge.exclamationmark",
            rarity: .epic,
            unlockCondition: "Log events for 30 days in a row"
        ) { baby, dataStore in
            await checkConsecutiveDays(baby: baby, dataStore: dataStore, requiredDays: 30)
        },

        // Age-based achievements
        Achievement(
            id: "first_month",
            title: "Month One Complete",
            description: "Your baby is one month old!",
            iconName: "moon.fill",
            rarity: .common,
            unlockCondition: "Reach 1 month of age"
        ) { baby, _ in
            let ageInDays = Calendar.current.dateComponents([.day], from: baby.dateOfBirth, to: Date()).day ?? 0
            return ageInDays >= 30
        },

        Achievement(
            id: "six_months",
            title: "Half Year Milestone",
            description: "Your baby is six months old!",
            iconName: "6.circle.fill",
            rarity: .rare,
            unlockCondition: "Reach 6 months of age"
        ) { baby, _ in
            let ageInDays = Calendar.current.dateComponents([.day], from: baby.dateOfBirth, to: Date()).day ?? 0
            return ageInDays >= 183
        },

        Achievement(
            id: "one_year",
            title: "First Birthday",
            description: "Your baby is one year old!",
            iconName: "birthday.cake.fill",
            rarity: .legendary,
            unlockCondition: "Reach 1 year of age"
        ) { baby, _ in
            let ageInDays = Calendar.current.dateComponents([.day], from: baby.dateOfBirth, to: Date()).day ?? 0
            return ageInDays >= 365
        },

        // Pattern achievements
        Achievement(
            id: "nap_predictor",
            title: "Nap Guru",
            description: "Used nap predictions for 30 days",
            iconName: "bed.double.fill",
            rarity: .rare,
            unlockCondition: "Use nap predictions for 30 days"
        ) { baby, dataStore in
            // This would need to track prediction usage - for now, check for consistent logging
            await checkConsecutiveDays(baby: baby, dataStore: dataStore, requiredDays: 30)
        },

        Achievement(
            id: "feeding_tracker",
            title: "Feeding Expert",
            description: "Logged 200 feeding events",
            iconName: "bottle.fill",
            rarity: .rare,
            unlockCondition: "Log 200 feeding events"
        ) { baby, dataStore in
            let events = try? await dataStore.fetchEvents(for: baby, from: Date.distantPast, to: Date.distantFuture)
            let feedEvents = events?.filter { $0.type == .feed } ?? []
            return feedEvents.count >= 200
        }
    ]

    // Helper function to check consecutive days of logging
    private static func checkConsecutiveDays(baby: Baby, dataStore: DataStore, requiredDays: Int) async -> Bool {
        do {
            let thirtyDaysAgo = Calendar.current.date(byAdding: .day, value: -requiredDays, to: Date())!
            let events = try await dataStore.fetchEvents(for: baby, from: thirtyDaysAgo, to: Date())

            // Group events by date
            let calendar = Calendar.current
            var eventDates = Set<Date>()

            for event in events {
                let date = calendar.startOfDay(for: event.startTime)
                eventDates.insert(date)
            }

            // Check if we have events for each of the last N days
            for dayOffset in 0..<requiredDays {
                let checkDate = calendar.date(byAdding: .day, value: -dayOffset, to: Date())!
                let startOfDay = calendar.startOfDay(for: checkDate)

                if !eventDates.contains(startOfDay) {
                    return false
                }
            }

            return true
        } catch {
            return false
        }
    }
}

File: src/services/achievementService.ts
import localforage from 'localforage';

const achievementStore = localforage.createInstance({
  name: 'nestling',
  storeName: 'achievements',
});

export interface Achievement {
  id: string;
  title: string;
  description: string;
  icon: string;
  unlockedAt?: string;
}

const ACHIEVEMENTS: Achievement[] = [
  {
    id: 'first_week',
    title: 'First Week',
    description: 'Logged events for 7 consecutive days',
    icon: '🎉',
  },
  {
    id: 'night_owl',
    title: 'Night Owl',
    description: 'Logged a feed between 2am-5am',
    icon: '🦉',
  },
  {
    id: 'consistent_logger',
    title: 'Consistent Logger',
    description: 'Maintained a 7-day logging streak',
    icon: '📊',
  },
  {
    id: 'data_champion',
    title: 'Data Champion',
    description: 'Maintained a 30-day logging streak',
    icon: '🏆',
  },
  {
    id: 'early_adopter',
    title: 'Early Adopter',
    description: 'Signed up and started logging',
    icon: '⭐',
  },
  {
    id: 'milestone_master',
    title: 'Milestone Master',
    description: 'Logged 5 milestones',
    icon: '🎯',
  },
];

class AchievementService {
  async getAllAchievements(): Promise<Achievement[]> {
    return ACHIEVEMENTS;
  }

  async getUnlockedAchievements(babyId: string): Promise<Achievement[]> {
    const unlocked = await achievementStore.getItem<string[]>(`unlocked_${babyId}`);
    if (!unlocked) return [];

    return ACHIEVEMENTS.filter(a => unlocked.includes(a.id)).map(a => ({
      ...a,
      unlockedAt: unlocked.includes(a.id) ? new Date().toISOString() : undefined,
    }));
  }

  async unlockAchievement(babyId: string, achievementId: string): Promise<boolean> {
    const unlocked = (await achievementStore.getItem<string[]>(`unlocked_${babyId}`)) || [];

    if (unlocked.includes(achievementId)) {
      return false; // Already unlocked
    }

    unlocked.push(achievementId);
    await achievementStore.setItem(`unlocked_${babyId}`, unlocked);
    return true;
  }

  async checkAndUnlockAchievements(
    babyId: string,
    context: {
      streakDays?: number;
      eventType?: string;
      eventTime?: Date;
      milestoneCount?: number;
    }
  ): Promise<Achievement[]> {
    const newlyUnlocked: Achievement[] = [];

    // Check streak-based achievements
    if (context.streakDays) {
      if (context.streakDays >= 7) {
        const wasNew = await this.unlockAchievement(babyId, 'first_week');
        if (wasNew) {
          newlyUnlocked.push(ACHIEVEMENTS.find(a => a.id === 'first_week')!);
        }
      }
      if (context.streakDays >= 7) {
        const wasNew = await this.unlockAchievement(babyId, 'consistent_logger');
        if (wasNew) {
          newlyUnlocked.push(ACHIEVEMENTS.find(a => a.id === 'consistent_logger')!);
        }
      }
      if (context.streakDays >= 30) {
        const wasNew = await this.unlockAchievement(babyId, 'data_champion');
        if (wasNew) {
          newlyUnlocked.push(ACHIEVEMENTS.find(a => a.id === 'data_champion')!);
        }
      }
    }

    // Check night owl
    if (context.eventType === 'feed' && context.eventTime) {
      const hour = context.eventTime.getHours();
      if (hour >= 2 && hour < 5) {
        const wasNew = await this.unlockAchievement(babyId, 'night_owl');
        if (wasNew) {
          newlyUnlocked.push(ACHIEVEMENTS.find(a => a.id === 'night_owl')!);
        }
      }
    }

    // Check milestone master
    if (context.milestoneCount && context.milestoneCount >= 5) {
      const wasNew = await this.unlockAchievement(babyId, 'milestone_master');
      if (wasNew) {
        newlyUnlocked.push(ACHIEVEMENTS.find(a => a.id === 'milestone_master')!);
      }
    }

    return newlyUnlocked;
  }
}

export const achievementService = new AchievementService();
